### PR TITLE
extending afalg with aes-cbc-192/256, afalgtest.c also updated accord…

### PR DESCRIPTION
**Pull request by lullajd**
HI 
Please consider the new pull request which considers the latest comments from matt and stephen.
The changes are for extending afalg support for aes cbc key sizes 192 and 256 bits.
The tests also have been added which encrypt and match the encrypted bytes with known good values. Also the tests decrypt and match  the decrypted bytes with the original plain text.

The known good encrypted bytes have been taken as the following (without afalg engine):

jlulla@ubuntu:~/openssl-master2/apps$ od --format=x1 /home/jlulla/plain
0000000 53 69 6e 67 6c 65 20 62 6c 6f 63 6b 20 6d 73 67
0000020 0a

jlulla@ubuntu:~/openssl-master2/apps$ ./openssl enc -aes128 -in /home/jlulla/plain -out /home/jlulla/encrypted1  -K 06a9214036b8a15b512e03d534120006 -iv 3dafba429d9eb430b422da802c9fac41

jlulla@ubuntu:~/openssl-master2/apps$ od --format=x1 /home/jlulla/encrypted1    
0000000 e3 53 77 9c 10 79 ae b8 27 08 94 2d be 77 18 1a
0000020 2d 34 aa 3e 5a 6f 8d 05 69 ef 36 5d 48 27 a9 bb

jlulla@ubuntu:~/openssl-master2/apps$ ./openssl enc -aes192 -in /home/jlulla/plain -out /home/jlulla/encrypted1  -K 06a9214036b8a15b512e03d53412000606a9214036b8a15b -iv 3dafba429d9eb430b422da802c9fac41

jlulla@ubuntu:~/openssl-master2/apps$ od --format=x1 /home/jlulla/encrypted1    
0000000 f7 e4 26 d1 d5 4f 8f 39 b1 9e e0 df 61 b9 c2 55
0000020 eb 7d 5a a2 1a 66 1e da 99 e9 14 2c 6c 23 e7 7c

jlulla@ubuntu:~/openssl-master2/apps$ ./openssl enc -aes256 -in /home/jlulla/plain -out /home/jlulla/encrypted1  -K 06a9214036b8a15b512e03d53412000606a9214036b8a15b512e03d534120006 -iv 3dafba429d9eb430b422da802c9fac41

jlulla@ubuntu:~/openssl-master2/apps$ od --format=x1 /home/jlulla/encrypted1    
0000000 a0 76 85 fd c1 65 71 9d c7 e9 13 6e ae 55 49 b4
0000020 13 aa 1d a7 39 f3 df c3 40 66 d0 7b dd d5 30 d5

Please review the changes. 

Thanks
Jitendra 

Previous pull requests:

https://github.com/openssl/openssl/pull/4448
https://github.com/openssl/openssl/pull/4465
https://github.com/openssl/openssl/pull/4482
https://github.com/openssl/openssl/pull/4484
https://github.com/openssl/openssl/pull/4567

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: engines/e_afalg.c [hunk 1]
```diff
@@ -637,29 +643,45 @@ static int afalg_cipher_cleanup(EVP_CIPHER_CTX *ctx)
...
-                                                  sizeof(afalg_ctx)))) {
-        EVP_CIPHER_meth_free(_hidden_aes_128_cbc);
-        _hidden_aes_128_cbc = NULL;
-    }
-    return _hidden_aes_128_cbc;
+    cbc_handles *cipher_handle = get_cipher_handle(nid);
+    if (cipher_handle->_hidden == NULL
+        && ((cipher_handle->_hidden =
+         EVP_CIPHER_meth_new(nid,
+                  AES_BLOCK_SIZE,
```
Comment (mattcaswell) [inline engines/e_afalg.c]: nit: The indentation here should align `AES_BLOCK_SIZE` with `nid`. Similar comments apply to most of this `if` block

File: test/afalgtest.c [hunk 2]
```diff
@@ -41,31 +41,63 @@ static ENGINE *e;
...
-    RAND_bytes(in, BUFFER_SIZE);
 
     if (!TEST_true(EVP_CipherInit_ex(ctx, cipher, e, key, iv, 1))
             || !TEST_true(EVP_CipherUpdate(ctx, ebuf, &encl, in, BUFFER_SIZE))
             || !TEST_true(EVP_CipherFinal_ex(ctx, ebuf+encl, &encf)))
         goto end;
     encl += encf;
 
+    if (TEST_mem_eq(enc_result, BUFFER_SIZE, ebuf, BUFFER_SIZE))
+        printf("Encryption test passed for keysize %d\n",
```
Comment (mattcaswell) [inline test/afalgtest.c]: nit: Don't printf. On success we don't care. On failure the TEST macros will print suitable diagnostic messages.

**Comment by lullajd** [comment]
Thanks Matt for approving it.
I have fixed the indentation and have removed printf from afalgtest.c.

HI
Please consider the pull request.
The changes are for extending afalg support for aes cbc key sizes 192 and 256 bits.
The tests also have been added which encrypt and match the encrypted bytes with known good values. Also the tests decrypt and match the decrypted bytes with the original plain text.

Thanks
Jitendra

**Comment by mattcaswell** [comment]
Darn, I was about to merge this but I'm getting some compile errors when I use the "--strict-warnings" Configure flag (why didn't travis pick this up?). Anyway looks like this needs to be fixed:

````
gcc  -Iinclude -DDSO_DLFCN -DHAVE_DLFCN_H -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-1.1\"" -Wall -O0 -g -pthread -m64 -DL_ENDIAN  -DDEBUG_UNUSED -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wswitch -Wsign-compare -Wmissing-prototypes -Wshadow -Wformat -Wtype-limits -Wundef -Werror -Wmisleading-indentation -Wa,--noexecstack -fPIC -DOPENSSL_USE_NODELETE -MMD -MF engines/e_afalg.d.tmp -MT engines/e_afalg.o -c -o engines/e_afalg.o engines/e_afalg.c
engines/e_afalg.c:646:14: error: no previous prototype for 'get_cipher_handle' [-Werror=missing-prototypes]
 cbc_handles *get_cipher_handle(int nid)
              ^~~~~~~~~~~~~~~~~
engines/e_afalg.c: In function 'free_cbc':
engines/e_afalg.c:849:18: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     for(i = 0; i < OSSL_NELEM(afalg_cipher_nids); i++) {
                  ^
cc1: all warnings being treated as errors
Makefile:6378: recipe for target 'engines/e_afalg.o' failed
make[1]: *** [engines/e_afalg.o] Error 1
make[1]: Leaving directory '/home/matt/dev/openssl-write'
Makefile:139: recipe for target 'all' failed
make: *** [all] Error 2
````

**Comment by lullajd** [comment]
Sorry for the trouble but I could not catch it on my system.

I will fix this shortly. 


--------------------------------------------
On Tue, 11/14/17, mattcaswell <notifications@github.com> wrote:

 Subject: Re: [openssl/openssl] extending afalg with aes-cbc-192/256, afalgtest.c also updated accord… (#4717)
 To: "openssl/openssl" <openssl@noreply.github.com>
 Cc: "Jitendra Lulla" <lullajd@yahoo.com>, "Author" <author@noreply.github.com>
 Date: Tuesday, November 14, 2017, 10:15 AM
 
 Darn, I was about to
 merge this but I'm getting some compile errors when I
 use the "--strict-warnings" Configure flag (why
 didn't travis pick this up?). Anyway looks like this
 needs to be fixed:
 gcc  -Iinclude -DDSO_DLFCN -DHAVE_DLFCN_H
 -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC
 -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT
 -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM
 -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM
 -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM
 -DPADLOCK_ASM -DPOLY1305_ASM
 -DOPENSSLDIR="\"/usr/local/ssl\""
 -DENGINESDIR="\"/usr/local/lib/engines-1.1\""
 -Wall -O0 -g -pthread -m64 -DL_ENDIAN  -DDEBUG_UNUSED
 -DPEDANTIC -pedantic -Wno-long-long -Wall -Wextra
 -Wno-unused-parameter -Wno-missing-field-initializers
 -Wswitch -Wsign-compare -Wmissing-prototypes -Wshadow
 -Wformat -Wtype-limits -Wundef -Werror
 -Wmisleading-indentation -Wa,--noexecstack -fPIC
 -DOPENSSL_USE_NODELETE -MMD -MF engines/e_afalg.d.tmp -MT
 engines/e_afalg.o -c -o engines/e_afalg.o engines/e_afalg.c
 engines/e_afalg.c:646:14: error: no previous prototype for
 'get_cipher_handle' [-Werror=missing-prototypes]
  cbc_handles *get_cipher_handle(int nid)
               ^~~~~~~~~~~~~~~~~
 engines/e_afalg.c: In function 'free_cbc':
 engines/e_afalg.c:849:18: error: comparison between signed
 and unsigned integer expressions [-Werror=sign-compare]
      for(i = 0; i < OSSL_NELEM(afalg_cipher_nids); i++) {
                   ^
 cc1: all warnings being treated as errors
 Makefile:6378: recipe for target 'engines/e_afalg.o'
 failed
 make[1]: *** [engines/e_afalg.o] Error 1
 make[1]: Leaving directory
 '/home/matt/dev/openssl-write'
 Makefile:139: recipe for target 'all' failed
 make: *** [all] Error 2
 
 
 —
 You are receiving this because you authored the
 thread.
 Reply to this
 email directly, view
 it on GitHub, or mute
 the thread.

**Comment by mattcaswell** [comment]
I think the error message around the `get_cipher_handle` problem is a bit misleading. Actually I think the problem is that it needs to be declared "static".

**Comment by lullajd** [comment]
Hi,
I have fixed the warnings coming when --strict-warnings was supplied to configure script.
Thanks
Jitendra

**Comment by lullajd** [comment]
As correctly suggested by Matt,  get_cipher_handle should be static. Making this change.

File: engines/e_afalg.c [hunk 3]
```diff
@@ -637,29 +644,45 @@ static int afalg_cipher_cleanup(EVP_CIPHER_CTX *ctx)
     return 1;
 }
 
-const EVP_CIPHER *afalg_aes_128_cbc(void)
+cbc_handles *get_cipher_handle(int nid)
```
Comment (mattcaswell) [inline engines/e_afalg.c]: nit: For consistency please add `static` here too.

**Comment by lullajd** [comment]
Added the missing static.

**Comment by lullajd** [comment]
Hi Pauli,
Could you please review and approve?
Thanks
Jitendra

**Comment by mattcaswell** [comment]
Ping @paulidale

**Comment by mattcaswell** [comment]
Ping @paulidale

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**